pkg/apis/workspaces/v1alpha2/attributes.go

```diff
@@ -0,0 +1,9 @@
+package v1alpha2
+
+import (
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+// Attributes provides a way to add a map of arbitrary YAML/JSON
+// objects.
+type Attributes map[string]apiext.JSON
```

Comment:
from the model point of view, it's so useful to have any JSON in the field, but from client(odo, devworkspace) perspective of view, it means that schema does not validate attributes format and every field access requires checking that attributes have the right format, if I did not miss something. So, we need some library code that allows easily to get attribute value as string, int, interface or error if the field has the wrong format.
Am I missing something and it's not going to be an issue for clients to read the attributes?

Reply:
You're right, I'm looking into it.

Reply:
To add here, I don't know how elegant or useful embedding full json as an attribute will ultimately be. Even with `map[string]string` you _could_ ostensibly use json, but the main problem is that it's hard to write. I think I'd rather see something like
```yaml
attributes:
  controller.devfile.io/plugin/name: "vscode-java"
  controller.devfile.io/plugin/vscode-extensions/java: "https://my-url.com/java.vsix"
  controller.devfile.io/plugin/vscode-extensions/java-debug: "https://my-url.com/java-debug.vsix"
```
than have the above collapsed into a json blob
```yaml
attributes:
  controller.devfile.io/plugin: '{"name": "vscode-java", "plugins": ["java.vsix", "java-debug.vsix"]}'
```
The latter could get quite complicated as there are longer entries. It also forces attributes to be complex, as we couldn't use something like `my-attribute: my-value`.

Reply:
> than have the above collapsed into a json blob

@amisevsk we just don't need to have json collasped into a JSON blob.
This is a json-like arbitrary value, directly included inside the JSON or YAML document.
So in a YAML document it will a yaml fragment, and in a JSON document it will be a JSON document.

Your example would be:

```yaml
    attributes:
      name: vscode-java
      plugins:
        - java.vsix
        - java-debug.vsix
```

That's the whole point of the change.

Reply:
And with the helpers I'm currently adding to the PR, when getting the GO structs, you would extract it like that:

```go
var Attributes attributes = ...
myName, isAString := attributes.Get("name").Interface().(string)

// And for plugins you might use want to use the `DecodeInto` method, to decode the arbitrary JSON-like structure into a given, well-known GO type. 

plugins := &[]string{}
err := attributes.Get("plugins").DecodeInto(plugins)

// Of course you also have a method to know if a given key is available in the Attributes:

isThere := attributes.Exists("aMissingKey")
```

Reply:
This makes sense to me and is a lot cleaner than my suggestion :)

Reply:
I am wondering if it makes sense to have these helper functions in devfile/api or devfile/library

Reply:
@maysunfaisal In fact it is there, because my plan was to also use it in the `DevfileMetadata` struct, which is also in the `devfile/api` repo, but in the non-K8S-versioned `devfile` package.